File: VideoStreamingBackend/domain/entities/stream_session.py
"""StreamSession domain entity."""
from __future__ import annotations

from dataclasses import dataclass, field
from datetime import datetime, timezone
from enum import Enum
from typing import Optional
from uuid import UUID, uuid4


class StreamStatus(str, Enum):
    """Status of a stream session."""
    PENDING = "pending"
    CONNECTING = "connecting"
    ACTIVE = "active"
    RECONNECTING = "reconnecting"
    STOPPED = "stopped"
    ERROR = "error"


@dataclass
class StreamSession:
    """
    Represents an active video streaming session for a camera.
    
    Attributes:
        id: Unique session identifier
        camera_id: ID of the camera being streamed
        stream_url: URL of the camera stream (RTSP/HTTP)
        status: Current session status
        fps: Frames per second
        started_at: When the session started
        last_frame_at: Timestamp of last received frame
        error_message: Error details if status is ERROR
        reconnect_attempts: Number of reconnection attempts
    """
    id: UUID
    camera_id: UUID
    stream_url: str
    status: StreamStatus = StreamStatus.PENDING
    fps: int = 15
    started_at: Optional[datetime] = None
    last_frame_at: Optional[datetime] = None
    stopped_at: Optional[datetime] = None
    error_message: Optional[str] = None
    reconnect_attempts: int = 0
    created_at: datetime = field(default_factory=lambda: datetime.now(timezone.utc))
    updated_at: datetime = field(default_factory=lambda: datetime.now(timezone.utc))

    @classmethod
    def create(
        cls,
        camera_id: UUID,
        stream_url: str,
        fps: int = 15,
    ) -> StreamSession:
        """Factory method to create a new stream session."""
        now = datetime.now(timezone.utc)
        return cls(
            id=uuid4(),
            camera_id=camera_id,
            stream_url=stream_url,
            fps=fps,
            status=StreamStatus.PENDING,
            created_at=now,
            updated_at=now,
        )

    def start(self) -> None:
        """Mark the session as started."""
        self.status = StreamStatus.CONNECTING
        self.started_at = datetime.now(timezone.utc)
        self.updated_at = datetime.now(timezone.utc)

    def activate(self) -> None:
        """Mark the session as active (receiving frames)."""
        self.status = StreamStatus.ACTIVE
        self.updated_at = datetime.now(timezone.utc)

    def stop(self) -> None:
        """Stop the session."""
        self.status = StreamStatus.STOPPED
        self.stopped_at = datetime.now(timezone.utc)
        self.updated_at = datetime.now(timezone.utc)

    def mark_error(self, message: str) -> None:
        """Mark the session as errored."""
        self.status = StreamStatus.ERROR
        self.error_message = message
        self.updated_at = datetime.now(timezone.utc)

    def mark_reconnecting(self) -> None:
        """Mark the session as reconnecting."""
        self.status = StreamStatus.RECONNECTING
        self.reconnect_attempts += 1
        self.updated_at = datetime.now(timezone.utc)

    def update_last_frame(self) -> None:
        """Update the last frame timestamp."""
        self.last_frame_at = datetime.now(timezone.utc)
        self.updated_at = datetime.now(timezone.utc)

    def is_alive(self, timeout_seconds: int = 30) -> bool:
        """Check if the session is still receiving frames."""
        if self.status not in (StreamStatus.ACTIVE, StreamStatus.RECONNECTING):
            return False
        if self.last_frame_at is None:
            return False
        elapsed = (datetime.now(timezone.utc) - self.last_frame_at).total_seconds()
        return elapsed < timeout_seconds

File: VideoStreamingBackend/src/domain/entities/StreamSession.ts
/**
 * Video Streaming Service — Domain Layer
 * Stream entity representing an active camera ingest + SFU routing session.
 *
 * In the WebRTC architecture:
 * - Each camera has ONE StreamSession (ingest via FFmpeg → RTP → mediasoup Producer)
 * - Multiple ViewerSessions consume from the same Producer
 * - The StreamManager orchestrates lifecycle; mediasoup handles media routing.
 */

export type StreamStatus =
  | 'pending'       // Created, not yet ingesting
  | 'connecting'    // FFmpeg starting / probing RTSP
  | 'active'        // RTP flowing, Producer receiving frames
  | 'reconnecting'  // FFmpeg died, auto-reconnect in progress
  | 'stopped'       // Intentionally stopped
  | 'error';        // Unrecoverable error

export interface StreamConfig {
  fps: number;
  width: number;
  height: number;
  codec: string;
  bitrate: number;
}

export interface StreamSession {
  /** Unique session ID (UUID) */
  id: string;
  /** Camera being streamed */
  cameraId: string;
  /** Camera owner's user ID */
  ownerUserId: string;
  /** RTSP/HTTP source URL */
  streamUrl: string;
  /** Current session status */
  status: StreamStatus;
  /** Stream configuration */
  config: StreamConfig;
  /** Number of active WebRTC consumers */
  viewerCount: number;

  // ─── mediasoup references (set after ingest starts) ───
  /** mediasoup Router ID for this camera */
  routerId: string | null;
  /** mediasoup video Producer ID */
  videoProducerId: string | null;
  /** mediasoup audio Producer ID (null if camera has no audio) */
  audioProducerId: string | null;

  // ─── Timestamps ───
  startedAt: Date | null;
  lastFrameAt: Date | null;
  stoppedAt: Date | null;

  // ─── Error / reconnect tracking ───
  errorMessage: string | null;
  reconnectAttempts: number;

  // ─── Metadata ───
  createdAt: Date;
  updatedAt: Date;
}

export const DEFAULT_STREAM_CONFIG: StreamConfig = {
  fps: 15,
  width: 1280,
  height: 720,
  codec: 'h264',
  bitrate: 2_000_000,
};
